Sources/NIOHTTPResponsiveness/HTTPDrippingDownloadHandler.swift

```diff
@@ -0,0 +1,246 @@
+        if case .dripping = self.phase, self.pendingWrite && context.channel.isWritable {
+            self.writeChunk(context: context)
+        }
+    }
+
+    private func writeChunk(context: ChannelHandlerContext) {
+        // Make sure we don't accidentally reenter
+        if self.activelyWritingChunk {
+            return
+        }
+        self.activelyWritingChunk = true
+        defer {
+            self.activelyWritingChunk = false
+        }
+
+        // If we're not dripping the response body, there's nothing to do
+        guard case .dripping(var drippingState) = self.phase else {
+            return
+        }
+
+        // If we've sent all chunks, send end and be done
+        if drippingState.chunksLeft < 1 {
+            self.phase = .done
+            context.writeAndFlush(self.wrapOutboundOut(.end(nil)), promise: nil)
+            return
+        }
+
+        var dataWritten = false
+        while drippingState.currentChunkBytesLeft > 0, context.channel.isWritable {
+            let toSend = min(drippingState.currentChunkBytesLeft, HTTPDrippingDownloadHandler.downloadBodyChunk.readableBytes)
+            let buffer = HTTPDrippingDownloadHandler.downloadBodyChunk.getSlice(
+                at: HTTPDrippingDownloadHandler.downloadBodyChunk.readerIndex,
+                length: toSend
+            )!
+            context.write(self.wrapOutboundOut(.body(buffer)), promise: nil)
+            drippingState.currentChunkBytesLeft -= toSend
+            dataWritten = true
+        }
+
+        // If we weren't able to send the full chunk, it's because the channel isn't writable. We yield until it is
+        if drippingState.currentChunkBytesLeft > 0 {
+            self.pendingWrite = true
+            self.phase = .dripping(drippingState)
+            if dataWritten {
+                context.flush()
+            }
+            return
+        }
+
+        // We sent the full chunk. If we have no more chunks to write, we're done!
+        drippingState.chunksLeft -= 1
+        if drippingState.chunksLeft == 0 {
+            self.phase = .done
+            context.writeAndFlush(self.wrapOutboundOut(.end(nil)), promise: nil)
+            return
+        }
+
+        if dataWritten {
+            context.flush()
+        }
+
+        // More chunks to write.. Kick off timer
+        drippingState.currentChunkBytesLeft = self.size
+        self.phase = .dripping(drippingState)
+        if self.scheduledCallbackHandler == nil {
+            let this = NIOLoopBound(self, eventLoop: context.eventLoop)
+            let loopBoundContext = NIOLoopBound(context, eventLoop: context.eventLoop)
+            self.scheduledCallbackHandler = HTTPDrippingDownloadHandlerScheduledCallbackHandler(handler: this, context: loopBoundContext)
+        }
+        // SAFTEY: scheduling the callback only potentially throws when invoked off eventloop
+        try! context.eventLoop.scheduleCallback(in: self.frequency, handler: self.scheduledCallbackHandler!)
```

Comment:
Different EL implementations could throw for other reasons. In practice that _probably_ won't be the case but it'd be safer to catch the error and call `context.fireErrorCaught` instead of `try!` here.

Reply:
Good point - fixed